### PR TITLE
backfill commit in source code without redo calculate

### PR DESCRIPTION
**Pull request by zwpaper**
main branch, indexed without commit:
CleanShot 2024-12-19 at 19 47 36@2x

backfill:
CleanShot 2024-12-19 at 20 18 14@2x

### CODE REVIEW

File: crates/tabby-index/src/code/index.rs [hunk 1]
```diff
@@ -165,6 +172,48 @@ fn is_valid_file(file: &SourceCode) -> bool {
...
+    let schema = tabby_common::index::IndexSchema::instance();
+    doc! {
+        schema.field_id => get_text(doc, schema.field_id),
+        schema.field_source_id => get_text(doc, schema.field_source_id).to_string(),
+        schema.field_corpus => get_text(doc, schema.field_corpus).to_string(),
+        schema.field_attributes => json!({
+            code::fields::ATTRIBUTE_COMMIT: commit,
+        }),
+        schema.field_updated_at => get_text(doc, schema.field_updated_at).to_string(),,
+        schema.field_failed_chunks_count => get_json_number_field(doc, schema.field_failed_chunks_count, code::fields::FAILED_CHUNKS_COUNT) as usize,
```
Comment (wsxiaoys) [inline crates/tabby-index/src/code/index.rs:185]: https://github.com/TabbyML/tabby/blob/c8735f5690ef82fd36afd38597eb966c207861df/crates/tabby-common/src/index/mod.rs#L64

field_failed_chunks_count is not a json field.

File: crates/tabby-common/src/index/mod.rs [hunk 2]
```diff
@@ -178,6 +178,21 @@ impl IndexSchema {
         ])
     }
 
+    /// Build a query to find the document with the given `doc_id`, include chunks.
```
Comment (wsxiaoys) [inline crates/tabby-common/src/index/mod.rs:181]: why move order?

File: crates/tabby-index/src/code/index.rs [hunk 3]
```diff
@@ -165,6 +172,48 @@ fn is_valid_file(file: &SourceCode) -> bool {
         && file.number_fraction <= MAX_NUMBER_FRACTION
 }
 
+fn create_document_with_commit(doc: &TantivyDocument, commit: &str) -> TantivyDocument {
```
Comment (wsxiaoys) [inline crates/tabby-index/src/code/index.rs:175]: can this method being somewhat refactored into `TantivyDocBuilder`? e.g `TantivyDocBuilder::backfill_doc_attribute`

File: crates/tabby-common/src/index/mod.rs (hunk 2 above)
Comment (zwpaper) [reply crates/tabby-common/src/index/mod.rs:181]: The `doc_query_with_chunks` function should remain adjacent to the `doc_query` function and not be separated by `doc_indexed_after` and other functions.

File: crates/tabby-index/src/code/index.rs [hunk 4]
```diff
@@ -101,7 +102,19 @@ async fn add_changed_documents(
 
             let id = SourceCode::to_index_id(&repository.source_id, &key).id;
 
+            // Skip if already indexed and has no failed chunks,
+            // when skip, we should check if the document needs to be backfilled.
             if !require_updates(cloned_index.clone(), &id) {
+                backfill_commit_if_needed(
```
Comment (wsxiaoys) [inline crates/tabby-index/src/code/index.rs:108]:
```suggestion
                backfill_commit_in_doc_if_needed(
```

File: crates/tabby-index/src/indexer.rs [hunk 5]
```diff
@@ -167,6 +167,22 @@ impl<T: ToIndexId> TantivyDocBuilder<T> {
             }
         }
     }
+
+    pub async fn backfill_attribute(&self, origin: &TantivyDocument, doc: &T) -> TantivyDocument {
```
Comment (wsxiaoys) [inline crates/tabby-index/src/indexer.rs:171]:
```suggestion
    pub async fn backfill_doc_attributes(&self, origin: &TantivyDocument, doc: &T) -> TantivyDocument {
```

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

**Comment by wsxiaoys** [comment]
Otherwise LGTM

**Pull request merged**